Tasks/DeployVisualStudioTestAgent/DeployTestAgent.ps1

```diff
@@ -1,111 +1,117 @@
+    $TestAgentConfigurationScript       = "$PSScriptRoot\TestAgentConfiguration.ps1"
+    $testAgentHelperScript              = "$PSScriptRoot\TestAgentHelper.ps1"
+    $installTestAgentScript             = "$PSScriptRoot\InstallTestAgent.ps1"
+    $verifyTestAgentInstalledScript     = "$PSScriptRoot\VerifyTestAgentInstalled.ps1"
+
+    # Fix Assembly Redirections
+    # VSTS uses Newton Json 8.0 while the System.Net.Http uses 6.0
+    # Redirection to Newton Json 8.0
+    $jsonAssembly = [reflection.assembly]::LoadFrom($PSScriptRoot + "\modules\Newtonsoft.Json.dll") 
+    $onAssemblyResolve = [System.ResolveEventHandler] {
+        param($sender, $e)
+        if ($e.Name -eq "Newtonsoft.Json, Version=6.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed") { return $jsonAssembly }
+        foreach($a in [System.AppDomain]::CurrentDomain.GetAssemblies())
+        {
+            if($a.FullName -eq $e.Name) { return $a } else { return $null }
+        }
+        return $null
+    }
 
-if (!$personalAccessToken)
-{
-    Write-Host "##vso[task.logissue type=error;code=Unable to generate Personal Access Token for the user;TaskName=DTA]"
-    throw (Get-LocalizedString -Key "Unable to generate Personal Access Token for the user. Contact Project Collection Administrator")
-}
+    # Import the Task.Internal dll that has all the cmdlets we need for Build
+    Import-Module "$PSScriptRoot\modules\Microsoft.TeamFoundation.DistributedTask.Task.Deployment.dll"
+    Import-Module "$PSScriptRoot\modules\MS.TF.Task.TestPlatform.Acquisition.dll"
+    
+    Write-Verbose "Getting Access Token for the Run"
+    $endpoint = Get-VstsEndpoint -Name SystemVssConnection -Require
+    $personalAccessToken = [string]($endpoint.auth.parameters.AccessToken)
 
-try
-{
-    $persistExists = CmdletHasMember "Persist"
+    if (!$personalAccessToken)
+    {
+        Write-Host "##vso[task.logissue type=error;code=Unable to generate Personal Access Token for the user;TaskName=DTA]"
+        throw (Get-LocalizedString -Key "Unable to generate Personal Access Token for the user. Contact Project Collection Administrator")
+    }
 
-    if($persistExists)
+    try
     {
-        Write-Verbose "Calling Register Environment cmdlet"
-        $environment = Register-Environment -EnvironmentName $testMachineGroup -EnvironmentSpecification $testMachineGroup -UserName $adminUserName -Password $adminPassword -TestCertificate ($testCertificate -eq "true") -Connection $connection -TaskContext $distributedTaskContext -WinRmProtocol $winRmProtocol -ResourceFilter $testMachines -Persist
-        Write-Verbose "Environment details $environment"
+        Write-Verbose "Completed Register-Environment" -Verbose
+        $environment = Register-Environment -EnvironmentName $testMachineGroup -EnvironmentSpecification $testMachineGroup -ResourceFilter $testMachines -WinRmProtocol $winRmProtocol -TestCertificate ($testCertificate -eq "true") -UserName $adminUserName -Password $adminPassword
```

Comment:
all the tasks are retried 3 times by default. any failures, respective telemetry ll be logged from DLL